include/dg/llvm/ValueRelations/AnalysisGraph.hpp

```diff
@@ -0,0 +1,198 @@
+
+        relations.ddump(); // mark parameter const when deleting
+
+        const AllocatedArea* area = nullptr;
+        unsigned index = 0;
+
+        for (const llvm::Value* equal : relations.getEqual(gep->getPointerOperand())) {
+            std::tie(index, area) = structure.getAllocatedAreaFor(equal);
+            if (area) break;
+        }
+        if (! area) return "maybe"; // memory was not allocated by ordinary means (or at all)
+        if (! validMemory[index]) return "maybe"; // memory is not valid at given location
+
+        const std::vector<AllocatedSizeView>& views = area->getAllocatedSizeViews();
+        if (views.empty()) return "unknown"; // the size of allocated memory cannot be determined
+
+        if (gep->hasAllZeroIndices()) return readSize <= views[0].elementSize ? "true" : "maybe";
+        // maybe, because can read i64 from an array of two i32
+
+        const llvm::Value* gepIndex;
+        const llvm::Type* gepType;
+        std::tie(gepIndex, gepType) = getOnlyNonzeroIndex(gep);
+        // if gep has more indices than one, or there are zeros after
+        if (! gepIndex) return "unknown";
+
+        uint64_t gepElem = getBytes(gepType);
+
+        //std::cerr << "[readSize] " << readSize << std::endl;
+        //std::cerr << "[allocElem] " << allocElem << std::endl;
+        //std::cerr << "[gepElem] " << gepElem << std::endl;
+        //std::cerr << "[allocCount] " << dg::debug::getValName(allocCount) << std::endl;
+        //std::cerr << "[gepIndex] " << dg::debug::getValName(gepIndex) << std::endl;
+
+        // DANGER just an arbitrary type
+        llvm::Type* i32 = llvm::Type::getInt32Ty(views[0].elementCount->getContext());
+        const llvm::Constant* zero = llvm::ConstantInt::getSigned(i32, 0);
+
+        // check if index doesnt point before memory
+        if (relations.isLesser(gepIndex, zero))
+            // we know that pointed memory is alloca because allocCount is not nullptr
+            return "false";
+        if (! relations.isLesserEqual(zero, gepIndex)) return "maybe";
+
+        // check if index doesnt point after memory
+        for (const AllocatedSizeView& view : views) {
+
+            std::cerr << "inloop" << std::endl;
+            std::cerr << "[allocCount] " << dg::debug::getValName(allocCount) << std::endl;
+            std::cerr << "[gepIndex] " << dg::debug::getValName(gepIndex) << std::endl;
+            if (relations.isLesser(gepIndex, view.elementCount)) {
+                if (gepElem <= view.elementSize) return readSize <= view.elementSize ? "true" : "maybe";
+            }
+            if (relations.isLesserEqual(allocCount, gepIndex) && gepElem >= allocElem) return "false";
+
+            std::tie(allocCount, allocElem) = stripArithmeticOp(allocCount, allocElem);
+        } while (allocCount);
+
+        return "maybe";
+    }  
+
+public:
+    std::string isValidPointer(const llvm::Value* ptr, const llvm::Value* size) const {
```

Comment:
I suppose this method should be in the sbt-instrumentation plugin, not here.

Reply:
It is, in fact, there too. I developed it here because it was much more convinient for me. I will remove the file from the merge.